### DISCUSSION TITLE
iPXE Diskless Windows 10 - Can Install via to ISCSI via UEFI but not Legacy BIOS

### DISCUSSION THREAD

**Discussion by Fibbicles** [ipxe/ipxe#955] [General]
EDIT: Updated title to reflect current issue, please see my updated comment below.

So I recently moved away from ggRock and ccBoot after having so many issues with their user interfaces and decided to just build our own.

We currently have a Debian 11 install running:

- TGT
- Nginx
- Dnsmasq serving our ipxe.efi file
- Django application for handling pxe scripts
- ZFS for snapshot and volume management to deploy updates easily across a single image and send those snapshots to multiple machines

Now I've got a lot of internal code managing ZFS and will rundown how it works, iPXE.efi is using a custom script to request a pxe script from our Django application which will generate a snapshot of the desired base image for that machine, map it as an ISCSI target through tgt and then sanboot from ipxe.

To get the base image we attach the zfs volume as its own dedicated iscsi and shadow volume copy the entire base machine drive to the zfs volume and I have confirmed all files, drivers and registry hive settings are intact and relocated properly.

The problem is the dreaded inaccessible boot device error on Windows 10 after a while on the boot screen.
Here's everything I've tried:

- set gateway 0.0.0.0 in iPXE
- Tried both super image from ccboot and direct clone from local hdd on client system
- Made NIC driver service start type automatic instead of manual via regedit
- booted windows 10 installation media to install to iscsi, makes it to step 2 but hits same error as cloned images
- Also tried enabling boot log but it seems it never reaches a point to start logging

All of these computers ran fine on our old ggRock installation so I know BIOS settings and UEFI boot mode is all correct. Honestly willing to send a paypal reward to whoever can fix this as I've tried everything with no luck and I've got to get this clients online in 4 days for some classes using them.

**Comment by NiKiZe** [comment]
See several other discussions and issues opened in the repo about Win10 iSCSI issues,
Where these images running from iSCSI before? Is the NICs in the same slot, and with known working iSCSI drivers?

Some builds of Win10 seems to be no go, You might try to debug issues on which data is read and sent last, there is also packet filtering.
For example, does the NIC come up with the kernel driver at all, does it have any IP etc at the time of your bluescreen? (this is just general things you can look for, but I would recommend that you continue this in any of the already existing threads on the topic)

**Reply by Fibbicles** [reply]
> See several other discussions and issues opened in the repo about Win10 iSCSI issues, Where these images running from iSCSI before? Is the NICs in the same slot, and with known working iSCSI drivers?
> 
> Some builds of Win10 seems to be no go, You might try to debug issues on which data is read and sent last, there is also packet filtering. For example, does the NIC come up with the kernel driver at all, does it have any IP etc at the time of your bluescreen? (this is just general things you can look for, but I would recommend that you continue this in any of the already existing threads on the topic)

Everything is standard, yes it has an IP and I've essentially cloned the entire setup we previously had without the commercial management software on top because underneath it this exact setup. The weird thing is I know people mentioned issues with newer versions of Windows 10 but we had 22H2 running perfectly fine from ggRock so I suspect there's possibly something in their NIC driver collector that they're keeping secret that could fix so many problems for the community as I've had these exact same issues with ccBoot in the past also.

**Reply by NiKiZe** [reply]
Does it have an IP when the kernel uses the windows native driver? And not just when in PXE mode?

**Comment by stappersg** [comment]
> All of these computers ran fine on our old ggRock installation so I know BIOS settings and UEFI boot mode is all correct.

* Make a _wire capture_ with the `ggRock` setup
* Make a _wire capture_ with the new setup
* Compare the `.pcap` files
* Reduce the differences by changing the new server setup
* Make a new _wire capture_ with the newer setup
* Reduce further differences by further tuning the new server

> Honestly willing to send a paypal reward to whoever can fix this

I'm looking forward how the above advice is valued, email address in my github.com profile.

**Reply by Fibbicles** [reply]
Unfortunately we can't, ggRock was a hell of a thing to get working and licenses are expensive on top of having to go through 2-3 days of support any time you want to renew, their customer service is one of the main reasons we are leaving and I personally don't have much knowledge on network capturing.

We have this run on a proxmox installation with another VM set for IPXE booting to test other than our physical clients so if you want a login to see if you might be able to have a look by all means.

**Reply by stappersg** [reply]
Acknowledge on the _we can't_   and the further information that was in the response.

**Comment by Fibbicles** [comment]
Sorry quick update, I have got the installation successfully working.

The issue was very specific to the e1i64x64 driver which is used for a few different Intel Controllers, anyway by installing directly from Windows 10 setup files via PE I was able to install and successfully boot Windows 10. The problem I found was that Windows update then tried to install what I can only assume was also the latest version of the driver I was previously trying to inject into my images with which again lead to another "INACCESSIBLE_BOOT_DEVICE" error so I simply rolled back the driver to the one installed with the Windows 10 installation and now my setups work fine.

My only issue I am working through now is generating multiple system images as we have 3 different sets of clients we use and 2 can use UEFI fine, one however are old HP Compaq 8200 Elite which have EFI boot sources for all but the network adapter meaning it's the only one we have to run in Legacy mode, but the problem is during the install we get:
"Windows cannot be installed to this disk. The computer's hardware may not support booting to this disk. Ensure that the disk's controller is enabled in the computer's BIOS menu."

That message is only on Legacy BIOS as I tested it on our 2 other types of clients we use UEFI boot and had the same issue when on Legacy BIOS for the PXE boot.

**Reply by psychrabi** [reply]
When you initialized the target image in the server did you initialize as GPT or MBR ?
make sure its MBR for legacy boot support.
You can make image in MBR partitioned image and later add support for UEFI by creating a 100M fat32 partition
and copying the boot file using BCDboot C:\Windows /s D: /F all

D: is 100M fat32 partition
